src/main/clojure/clara/rules/engine.cljc

```diff
@@ -2002,6 +2008,29 @@
                  get-alphas-fn
                  []))
 
+(defn with-listener
+  [session listener]
+  (let [{:keys [listeners] :as components} (components session)]
+    (assemble (assoc components
+                     :listeners
+                     (conj listeners
+                           listener)))))
+
+(defn remove-listeners
+  [session pred]
+  (let [{:keys [listeners] :as components} (components session)
+        matching-listeners (filterv pred listeners)]
+    (if (seq matching-listeners)
+      (assemble (assoc components
+                       :listeners
+                       (remove pred listeners)))
```

Comment:
This seems like something that'd be better off not lazy. It just seems questionable when it will be realized later.
`(into [] (remove pred) listeners)`